x-pack/plugins/security_solution/server/lib/risk_engine/utils/create_ds.ts

```diff
@@ -0,0 +1,214 @@
+  const { index, alias } = concreteIndexInfo;
+  let simulatedIndexMapping: IndicesSimulateIndexTemplateResponse;
+  try {
+    simulatedIndexMapping = await retryTransientEsErrors(
+      () => esClient.indices.simulateIndexTemplate({ name: index }),
+      { logger }
+    );
+  } catch (err) {
+    logger.error(
+      `Ignored PUT mappings for alias ${alias}; error generating simulated mappings: ${err.message}`
+    );
+    return;
+  }
+
+  const simulatedMapping = get(simulatedIndexMapping, ['template', 'mappings']);
+
+  if (simulatedMapping == null) {
+    logger.error(`Ignored PUT mappings for alias ${alias}; simulated mappings were empty`);
+    return;
+  }
+
+  try {
+    await retryTransientEsErrors(
+      () => esClient.indices.putMapping({ index, body: simulatedMapping }),
+      { logger }
+    );
+
+    return;
+  } catch (err) {
+    logger.error(`Failed to PUT mapping for alias ${alias}: ${err.message}`);
+    throw err;
+  }
+};
+/**
+ * Updates the underlying mapping for any existing concrete indices
+ */
+const updateIndexMappings = async ({
+  logger,
+  esClient,
+  totalFieldsLimit,
+  concreteIndices,
+}: UpdateIndexMappingsOpts) => {
+  logger.debug(`Updating underlying mappings for ${concreteIndices.length} indices.`);
+
+  // Update total field limit setting of found indices
+  // Other index setting changes are not updated at this time
+  await Promise.all(
+    concreteIndices.map((index) =>
+      updateTotalFieldLimitSetting({ logger, esClient, totalFieldsLimit, concreteIndexInfo: index })
+    )
+  );
+
+  // Update mappings of the found indices.
+  await Promise.all(
+    concreteIndices.map((index) =>
+      updateUnderlyingMapping({ logger, esClient, totalFieldsLimit, concreteIndexInfo: index })
+    )
+  );
+};
+
+interface CreateConcreteWriteIndexOpts {
+  logger: Logger;
+  esClient: ElasticsearchClient;
+  totalFieldsLimit: number;
+  indexPatterns: IIndexPatternString;
+}
+/**
+ * Installs index template that uses installed component template
+ * Prior to installation, simulates the installation to check for possible
+ * conflicts. Simulate should return an empty mapping if a template
+ * conflicts with an already installed template.
+ */
+export const createConcreteWriteIndex = async ({
+  logger,
+  esClient,
+  indexPatterns,
+  totalFieldsLimit,
+}: CreateConcreteWriteIndexOpts) => {
+  logger.info(`Creating data stream - ${indexPatterns.alias}`);
+
+  // check if a concrete write index already exists
+  let concreteIndices: ConcreteIndexInfo[] = [];
+  try {
+    // Specify both the index pattern for the backing indices and their aliases
+    // The alias prevents the request from finding other namespaces that could match the -* pattern
+    const response = await retryTransientEsErrors(
+      () => esClient.indices.getDataStream({ name: indexPatterns.alias, expand_wildcards: 'all' }),
+      { logger }
+    );
+
+    concreteIndices = response.data_streams.map((dataStream) => ({
```

Comment:
These aren't really concrete indices, right? They're the datastream names themselves?